### ISSUE TITLE
Extend overlay information

### ISSUE THREAD

**Issue by MagielBruntink** [CXWorld/CapFrameX#111]
It would be great if the overlay could show the following PresentMon data:
- MsUntilDisplayed and the approximated input latency
- Runtime,SyncInterval,PresentFlags,AllowsTearing,PresentMode

**Comment by DevTechProfile** [comment]
Thanks for the suggestions. Do you mean current values or something like average values based on a sliding window with a length of 30 seconds for example?

**Comment by MagielBruntink** [comment]
I think what would be useful is:
- sliding window average presented as a number (PresentMon console output has something similar)
- linegraph similar time the other nunerics available to the overlay

(Awesome tool, btw!)

**Comment by Taxxor90** [comment]
Runtime is already covered by RTSS itself, the FPS/Frametime groupname shows it.
Presentmon itself doesn't even have the possibility to show the correct runtime right now, it's either "D3D9" for DX9, "DXGI" for DX11 and DX12 or "Other" for Vulkan and OpenGL.  RTSS can differentiate between all 5.

Sliding windows and linegraphs would only make sense for MSUntilDisplayed and the input lag, the other values dont change, at least not throughout my record files and even if they change for some frames, it's nothing that would be noticable in a sliding window.

**Comment by RodroG-85** [comment]
Maybe 'PresentMode' has nothing to do with the following, but I think it would be interesting if it actully allow to differenciate between presentation modes in some way: Fullscreen exclusive (FSE), Fullscreen Optimizations (FSO), Borderless Fullscreen Windowed (FSB).

**Comment by DevTechProfile** [comment]
Runtime,SyncInterval,PresentFlags,AllowsTearing,PresentMode are very special metrics. Not many people should be interested in this. Also, these values can be difficult to display in real time because we have a certain sample interval to refresh the overlay. Therefore the current value cannot be displayed. I would have to come up with another strategy. 

But, I will definitely implement the input lag. This should be interesting for many users.

**Comment by RodroG-85** [comment]
Ok, thanks. It meakes sense.

How can I see the 'PresentMode' data per record file? .json files allow it?

**Comment by Taxxor90** [comment]
You just have to search the json for PresentMode, there are as many entries as frames

**Comment by RodroG-85** [comment]
> You just have to search the json for PresentMode, there are as many entries as frames, so you can't make it per record file because the presentmode can change throughout the record

Searching the json for PresentMode doesn't offer any results though.

**Comment by xcasxcursex** [comment]
> **EDIT 2:** Searching old OCAT's .csv for PresentMode offers results for PresentMode per frame, and value doesn't change throughout the record (for example, 'Hardware Composed: Independent Flip' is constant on a DX11/Full screen/FSO enabled/V-Sync OFF scenario).

It may not change, but it can change. One example that comes to mind is the battlefield 1 and 4 games, I know that under certain circumstances, the game will change its presentation mode if alt-tabbing out and back into the game. Sorry I can't be more specific in my example, it was some time ago that I tested and remedied this, so I can't recall the exact parameters.... But the point is, the presentation mode _can_ differ from frame-to-frame.

> But, I will definitely implement the input lag. This should be interesting for many users.

You're right, input lag is very interesting for many users, however, this application isn't actually measuring input latency. I have concerns that this could lead to a great deal of misinformation in the community as this application grows in popularity. TL;DR at the bottom if you're in a rush :)

In order to measure input latency, we would need to be aware of the timing of the input and output - input latency is a measure of the processing time between the input and the resultant output. 

Obviously, there are some unknown timings involved here. While the latency from the graphics pipeline to the display is generally constant and invariable (as it is a result of hardware limitations, ie, the display hardware latency) and thus it can be easily accounted for by applying a constant (as this app does nicely with the "Input lag offset"), and the latency of physical input (pressing a key/mouse button) to input buffers (eg USB polling) is generally within a fairly insignificant window (1-5ms) and cannot be remedied by any simple means (such as changing graphics settings in-game) so is also fairly accounted for by the constant 'offset'... 

There is a greater issue here, which is that presentmon (and therefore CX) is not aware of the input sampling (ie, reading of the input from the buffers to the game to generate the frame). This is a factor in input latency which has a great impact on the actual input latency, as discussed [over here](https://github.com/GameTechDev/PresentMon/issues/73#issuecomment-594686228) it is the reason why in-game limiters tend to have input latency (measured with high-speed cameras) an entire frame-time in difference depending on the framerate limiter in use - and in those cases, presentmon/CX will show the same input lag measurement, despite it being significantly different.

Now, while I'm sure that technically minded people such as yourselves understand all of this and you're specifically clear that the input lag is an "approximation", I think it is worth considering the end-user of an application like this, the non-technically minded person such as a gamer, is going to read "input lag" and not consider all of these finer details. I think it's worth renaming the "input lag" measurements to a more accurate naming convention, such as "graphics lag" or "processing lag" or "output lag" or something along those lines.... Perhaps you could come up with a better name, but the point is that it really shouldn't be called "input" lag when we know nothing of when the "input" occurred. We don't know when the key was pressed, we don't know when it was polled in hardware and read to buffers, and most significantly, we don't know when the buffers were read by the application(game).

So, **TL;DR** "Input lag" should be renamed because presentmon doesn't measure input.

**Comment by DevTechProfile** [comment]
> You're right, input lag is very interesting for many users, however, this application isn't actually measuring input latency.

You're also right. Input lag consists of two parts. On the one hand there is the delay within the render pipeline and on the other hand the peripheral delay (mouse + keyboard) has to be added up.

What CX does is to calculate an **approximation** of the pipeline part. The calculation shows a very high correlation. In order not to ignore the peripheral part, the user can set a constant offset for it.

What we need to do as the dev team is to **communicate this clearly**.

**Comment by xcasxcursex** [comment]
> What CX does is to calculate an **approximation** of the pipeline part. The calculation shows a very high correlation. In order not to ignore the peripheral part, the user can set a constant offset for it.

Well, the big problem here is there is a third part to consider. 

CX makes a very accurate measurement of the graphics pipeline.... Especially since we are aware of quite accurate offsets to apply as a result of high-speed camera testing (eg, Battle(non)sense might have an 8ms frametime and observe 35ms of input latency so we know that about 27ms for a very high-end setup is appropriate.), we have a result that is a very accurate approximation of the time from the beginning of the pipeline, to the actual display visible to the user. Even the input hardware can be well accounted for by the constant offset, as even when input peripherals have some variation, it is so small as to be insignificant.

The issue here is that the constant offset can adequately account for the roughly constant peripheral delays, but it doesn't account for programmatic differences in the application - what happens inbetween the input peripherals, and the graphics pipeline. The ordering of the input sampling, and concurrency of sampling input during GPU context or wait times, becomes quite significant here.


> What we need to do as the dev team is to **communicate this clearly**.

That's a good call. The trick is explaining rocket science to the masses ;) I think this is the kind of thing best explained visually - like the saying goes, a picture tells a thousand words. Much easier to explain the ordering and concurrency of physical input, input buffering, input sampling, frame preparation, frame processing, and frame display, if that chain is drawn out, than to use a bunch of big confusing nerd words.

Sadly my illustration skillset is about mspaint-level XD so I can't be of much help with that beyond rough sketches.... Do you know a graphics wizard? Perhaps the person who did your nice logo?


> What name do you suggest for the approximation?

Put simply, I would say "Output Lag" should be used in place of "Input lag". 

Assuming an appropriate offset, CX provides a considerably accurate measure of the latency between the beginning of the graphics pipeline and the actual display being visible to the user - this really is a very accurate measure of actual output latency. Using the word "output", even to the novice users, will imply that there is a relationship to input, so they will intuitively understand that this is related to input latency - but they will also intuitively understand that there is a real difference, or you'd have used "input" like they are familiar with.

**Comment by RodroG-85** [comment]
> You're also right. Input lag consists of two parts. On the one hand there is the delay within the render pipeline and on the other hand the peripheral delay (mouse/keyboard + display) has to be added up.
What CX does is to calculate an approximation of the pipeline part. The calculation shows a very high correlation. In order not to ignore the peripheral part, the user can set a constant offset for it.
What we need to do as the dev team is to communicate this clearly.

Agree with you. I consider the communication could be improved but the concept and related terms are clear, meaningful and easy understables for general public. 

> Put simply, I would say "Output Lag" should be used in place of "Input lag".

I my humble opinion, this would be even a worse term in terms of communication to general public. **Approximate** input-lag or latency, input-lag or latency **approximation** is clear enough. The CX's implementation and current probabilistic and expected software approach is adequate as it is.

There is no need to give more thought to the concept at this point. If you consider that there would be possible programming factors that add latency, then increase your off-set and explain the included factors that determine your value, and then perform you benchmarking or analysis. In the end, the offset would work just as a constant variable in the methodology of the analysis for your comparison purposes.

Terms like **approximate** and **expected** input-lag or latency are key to describe this **software approach** and are _totally meaningful terms_ to anyone who can read and understand with a minimum logical and knowledge skills on the issue. The difference is pretty clear with _button-to-pixel_ studies using high-speed cameras, based on expensive and specialized extra hardware, by the way; and which are more accurate and optimal in order to estimate input-lag or latency. Of course, I think the later would be clear and easy understable unless someone do not want to do so or want to misinterpret it intentionally for some stange and unknown reason.

**NOTE.** We are clearly derailing the Topic of this thread.

**Comment by xcasxcursex** [comment]
> **Approximate** input-lag or latency, input-lag or latency **approximation** is clear enough.

That's exactly the problem. It's being quite clear.... that it is something which it is not. This measurement actually has nothing to do with the user's input. It is not an input lag measurement; approximate, expected, or otherwise.

> Terms like **approximate** and **expected** input-lag or latency are key to describe this **software approach** and are _totally meaningful terms_ to anyone who can read and understand with a minimum logical and knowledge skills on the issue.

But this is the problem, we are describing that it is what it is not. It does not measure input latency. It does not measure input, at all. The person with minimal logical and knowledge skills will be misled into thinking that it is what it is not, because the name specifically says that it is what it is not. The fact that they are totally meaningful terms is exactly the problem here, because the terms provide meaning that is false.


> f you consider that there would be possible programming factors that add latency, then increase your off-set

Firstly, we may or may not be aware of the added latency. It will always be possible that it is there, so this means it is a guess. Secondly, it is not necessarily a constant, and thus cannot be accounted for by the offset. This added delay can significantly differ in length from frame-to-frame. The offset is entirely adequate in accounting for generally constant and known delays such as the display latency and input-to-buffer latency, but this 'third factor' we're discussing is not constant and is unknown, so it is beyond the scope of a constant offset.



> Terms like **approximate** and **expected** input-lag or latency are key to describe this **software approach** 

It is entirely possible, to measure **in software**, the expected input latency, but this requires monitoring the OS's kernel to observe the input buffering (eg using ETW to observe USB HID), application's calls to read the input buffers (eg a debugger), etc., and adding those measurements to the measurements we are taking now. While this would not account for delays in the peripherals (display lag, USB polling, peripheral firmware (eg debounce), etc) those factors are well accounted for by a constant offset.

The issue here is not that we are using software to make the measurement, rather than hardware. The issue is that presentmon does not measure input like this. The first 'input' presentmon is aware of, is the application 'inputting' the frame to the graphics pipeline. Before this happens, the timing of the user input is unknown to us. 


Calling this measurement "approximate input latency" is like measuring the second half of a lap of a racecar and calling that "approximate lap time". We don't really know what happened in the first half of the lap, so it's not really telling us what the lap time is. Sure, we can add an offset if we know what the first half lap time tends to be, but what if the car (frame) is stuck behind the pace car (FPS limiter) for part of the lap? Now the actual lap time is quite long but we assume it is normal. What if we consider concurrency where the car can simultaneously exist on two parts of the track at one moment in time? This sounds like magic but thanks to multithreading, it's a thing, so now our lap time is actually close to half a lap time, but still we assume it is normal. If the race fan does not understand this, and we call it "lap time" then they are going to think it's "lap time" because the terminology is misleading.... But if we called that "sector time" then we'd be entirely accurate.



> **NOTE.** We are clearly derailing the Topic of this thread.

Perhaps this conversation should move to a new issue... Although, since it's a simple matter of terminology, I don't expect a great deal of conversation should be required.... But things don't always turn out how we expect ;)

**Comment by Taxxor90** [comment]
> This measurement actually has nothing to do with the user's input. It is not an input lag measurement; approximate, expected, or otherwise.

I'd say it has. 
The input lag isn't just the time from the users input to the PC getting that input, it's the complete time from the users input to the frame reacting on that input being displayed on the monitor.

So our approximation covers half of this way while in adding the 5-10ms offset accounting for the time from a button press to the PC recognizing the input and the time for the monitor to refresh after getting the corresponding frame.

The only thing it doesn't account for is how different software handles the input, but that's why we have an upper and lower bound to get the best and the worst case.
Either way, when comparing different settings for a change in input lag behaviour, it doesn't matter if our "expected" value is a little lower or al litte higher than the actual value, because if upper and lower bounds fall by 10ms when enabeling a input lag reduction option for example, you don't know exactly if the average input lag fell from 60ms to 50ms(upper bound) or from 40ms to 30ms(lower bound), but you know that you've reduced your input lag by 10ms.

**Comment by xcasxcursex** [comment]
> The input lag isn't just the time from the users input to the PC getting that input, it's the complete time from the users input to the frame reacting on that input being displayed on the monitor.
> 
> So our approximation covers half of this way while in adding the 5-10ms offset accounting for the time from a button press to the PC recognizing the input and the time for the monitor to refresh after getting the corresponding frame. And this offset is a constant, FPS limiters or other driver related things do not change the latency of your keyboard or your monitor.

Agree, this is what I've said above.

But it's not input latency. As you've said, that is the "time from the users input....". That's where the term "input latency" takes it's name. It's a time starting at the user's input - and presentmon/CX does not read user input, even at the software level.


> we have an upper and lower bound to get the best and the worst case.

This does not account for delays in reading the input from the buffers to the application. 

For example let's say we have the following chain to represent input latency

1) User presses key
2) Keypress is read into input buffers
3) There is a 5 second delay (let's make it ridiculously long to demonstrate the point)
4) The game reads the input buffers
5) The game generates a frame based on the user input
6) The frame is processed by the CPU
7) The frame is processed by the GPU
8) The frame is sent to the monitor
9) The monitor refreshes and the user can see the frame

Now, timing of steps 1-4 is entirely unknown to us. We can assume with decent accuracy, the time taken for 1 and 2, so now only 3 and 4 are unknown. Presentmon measures the time taken by 6 and 7, while 8 and 9 are fairly constant and again can be assumed with decent accuracy. But 3,4,5 are still unknown, and unmeasured. In this scenario, CX will not have an upper bound of over 5 seconds. But that's what the input lag would be.

> it doesn't matter if our "expected" value is a little lower or al litte higher than the actual value, because if upper and lower bounds fall by 10ms when enabeling a input lag reduction option for example, you don't know exactly if the average input lag fell from 60ms to 50ms(upper bound) or from 40ms to 30ms(lower bound), but you know that you've reduced your input lag by 10ms.

Exactly... but using the silly example above, what you don't know is that you've reduced it by 10ms, from 40 to 30ms, out of a total of 5 seconds. Now, if you're only concerned with the difference between the two tests, that's entirely adequate, but you still haven't measured your input latency. The average user does not understand such distinction. They read input latency and think input latency. They see 60-50ms and 40-30ms and that's what they consider the input latency to be.

This becomes a real issue (as opposed to the hyperbole above of a 5 second delay) when looking at frametime limiters, as per the discussion linked above. In this scenario, the estimates provided by presentmon and the offset added by CX, result in the same input latency estimate whether using an external or a good internal frametime limiter - and in actual fact, they typically differ by roughly an entire frame-time. To use your analogy of enabling an input latency reducing option, what can happen here is that an option is enabled which does in fact reduce latency, but the estimation given by presentmon/CX shows that no reduction has been made. We see 0ms difference, but in actual fact, there might be 16.7ms. So, to use your words, you "know" that it had no effect - but it DID have an effect. 

This can go so far as to showing a greater estimated input latency, when in fact it has been reduced, in other words you "know" that the input latency is worse with the in-game limiter, when in fact it is better. You could for example make this happen by having an external limiter, and then setting the in-game limiter to be a lower FPS than the external limiter. The external limiter will now not cause any delay and is essentially doing nothing, but the frametimes in-game have increased as per the lower FPS cap, and consequently the input latency estimation will have increased, but because the game is no longer waiting for the present call to complete before reading input to prepare the next frame, the actual input latency is actually lower.

There are other examples of how this can be a problem, and a big one that's (at long last) becoming more common are game engines where the game simulation and the user input that controls it, run in an entirely separate thread to the graphics processing. In this case, the measurement of presentmon/CX is actually of the latency between the simulation and the display ('simulation lag'), but the delay between user input and the simulation, the part of input lag that's most relevant to the user, is extremely small, and not measured at all. Good examples of this [here](https://www.reddit.com/r/pcmasterrace/comments/90vd78/ever_wondered_why_framerates_way_above_monitor/)


Of course, if the end-user is tech savvy and already understands all of this, then that's fine. The tech-savvy nerd is not the problem here. The problem is the average end-user. The nerd knows the limitations of the method to estimate 'input latency' and understands how it can be misleading because it isn't measuring input latency at all, but just a significant part of the input latency. The nerd knows that what he "knows" is wrong. The average end-user just sees "input lag" and knows "smaller is better" and that's about the end of it. Good documentation that clearly explains all of this to the end-user is key here, as was said above, but much of this is beyond the understanding of the average end-user. The documentation starts with the name itself, and naming it 'input lag' is very misleading to those who don't know better.


CX is a great application that makes some very nerdy tools accessible to the average Joe Bloggs end-user, and that's great, and as it should, the application is becoming more and more popular.... But that's not without its pitfalls. Consider this scenario: The average end-user uses CX and sees input latency not changing regardless of the framerate limiter in use. Now, high-speed camera testing shows that the estimate is false. Do you want to try to explain to them that it wasn't actually measuring input latency, AFTER they decide that CX lied to them and is junk, or BEFORE? The problem here is not CX, it's the users' lack of understanding. CX never lied to them, they just didn't understand. But try telling the average person they had it wrong all along and it was never your fault..... How many internet arguments do you see ending with "huh, my bad, I was wrong"? No, better to make sure they know right from the start, that this isn't measuring input latency.

**Comment by MagielBruntink** [comment]
Great discussion! My two cents: you could name the MsUntilDisplayed "Presentation Latency"or "Presentation Lag", I think that's the most accurate term. The time it takes from the present call until a flip is registered. This is a non-trivial component of the button-to-pixel latency, depending on buffering by the graphics API and the GPU driver (render queue, flip queue, etc.). Can be anywhere from 0 frames to 3-4 frames worth of latency, in addition to other factors like already discussed.

**Comment by DevTechProfile** [comment]
@xcasxcursex I think you misunderstood the concept of the upper bound. Here's Jeffersons post: https://github.com/GameTechDev/PresentMon/issues/73#issuecomment-589874806

**Comment by xcasxcursex** [comment]
Yes, I understand that. See above my hyperbolic example of a 5 second delay reading the inputs. Even the upper bound makes certain assumptions, which, while reasonable, _may_ be false.

I think it's difficult to imagine this from the perspective of the uneducated end-user who does not understand all this. 

A very simple example:
User compares RTSS and in-game limiter
User sees same upper and lower bound
User thinks they are the same because the numbers are the same

This is what happened in @RodroG-85 reddit post comparing RTSS and the new Nvidia limiter. Thankfully, one user observing the results was aware of himself enough to ask the question, and thankfully, rodrog took the time to go back and add an extra explanation of why when CX says "input latency" it might be wrong. 

Even @Taxxor90 has fallen into this trap when he says that a measurement of 10ms difference means that we "know" that there is a 10ms difference. We don't, because even if the upper/lower bounds are correct (which is not guaranteed), we might have swapped from lower to upper bound and the difference would actually be 10ms +/- the difference between the bounds.

The idea here is to prevent the misunderstanding like that, from happening all the time. It is safe to call this "input lag" if we understand how it is related to input lag and especially how it is not related.... But for the end-user who can be safely assumed not to understand such technicalities, the term "input lag" becomes misleading. The fine details of this even managed to trick taxxor90! I didn't notice it at the time, either. What chance does average end-user have, if two software developers follow one another into the trap?

Just as we can understand the significance of this measurement with respect to the actual input lag, we can understand its insignificance and the fact that it is not measuring input lag because it is not measuring input. But if the measurement is labelled "input lag" then "input lag" is what the user will think. Meanwhile, if we call it "Presentation lag", then the people who understand, will still understand, but we have easily prevented the misunderstanding of the end-user, because they no longer bring with them the presumptions of what "input lag" means.

**Comment by DevTechProfile** [comment]
@xcasxcursex We will consider your comments. Your efforts have not been in vain. Thanks for that. But we ask for a little patience because other things have a higher priority.

**Comment by xcasxcursex** [comment]
Absolutely, I understand! Better to add new features, than spend time explaining the old features to people ;)

**Comment by Taxxor90** [comment]
Added "Render lag" to overlay entries, showing the approx. input lag value without the offset set on the sync page. Value as average over the last 2 seconds.
